### ISSUE TITLE
An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale

### ISSUE THREAD

**Issue by Yphy** [Yphy/Paper-Review#6]
## Introduction
Transformer의 인코더를 통해 이미지의 Self Attention 을 구한 아주 의미있는 논문이다.
물론 inductive bias 때문에 구글 내부 데이터셋으로만 당시 EfficientNet을 이겼다는 한계가 있지만 이후 DEIT 같은 모델에선 이를 극복했다. 논문 앞부분 패치임베딩을 제외하고는 기존 transformer 구조와 같다.

![v1](https://user-images.githubusercontent.com/47969237/141840424-cbe14b64-9e18-4e83-b25d-181fabd636f5.jpg)
**Figure1.**

이미지를 고정된 패치로 쪼개서 각각의 패치들을 postion embeddings 와 함께 linearly embedding하고 그 결과값의 sequence를 트랜스포머 인코더에 넣었다. 분류를 하기 위해서 extra learnable "classification token"을 시퀀스에 추가하였다.

## 3. Method

3.1 VISION TRANSFORMER (VIT)

Figure1에서 보다시피, 트랜스포머는 시퀀스 토큰의 1d 임베딩 값들을 인풋으로 한다. 이를 이미지에 적용하기 위해서 HxWxC인 이미지를 2D 패치로 flatten시켜준다. 원본 이미지의 해상도인 HXW 는 정사각형 패치 P로 나눠지며 패치의 개수는 N = HW/P^2 이 된다. 

트랜스포머는 모든 레이어에 동일한 D 크기의 latent vector를 받기 때문에, 패치 임베딩 값들을 linear projection을 통해서D 차원의 크기로 통일시킨다.

BERT의 클래스 토큰(cls token)과 마찬가지로, 임베딩 패치 시퀀스의 앞부분에 z^0_0 = {x_{class}}를 추가함으로써 output인 z^0_L 이 image representation인 y 가 되게끔 만들어준다. 이 z^0_L 에 classification head가 붙어서 MLP를 거쳐 최종적으로 클래스 분류가 된다.

Positional embedding은 이미지 패치에 위치 정보를 담아주는 역할을 하며 2D 포지션임베딩을 했을때 그다지 효과가 없어서 기존의 1D로 충분했다고 한다.

트랜스포머 인코더는 multihead self-attention(MSA)와 MLP 블럭으로 구성되며 각각의 블럭 전에 Layer normalization이 적용된다.

    class PatchEmbedding(nn.Module):
        def __init__(self, in_channels: int = 3, patch_size: int = 16, emb_size: int = 768, img_size: int = 224):
            self.patch_size = patch_size
            super().__init__()
            self.projection = nn.Sequential(
                nn.Conv2d(in_channels, emb_size, kernel_size=patch_size, stride=patch_size),
                Rearrange('b e (h) (w) -> b (h w) e'),
            )
				    #cls 토큰 ,positional embedding 추가
            self.cls_token = nn.Parameter(torch.randn(1,1, emb_size)) 
            self.positions = nn.Parameter(torch.randn((img_size // patch_size) **2 + 1, emb_size))

        def forward(self, x: Tensor) -> Tensor:
            b, _, _, _ = x.shape
            x = self.projection(x)
            cls_tokens = repeat(self.cls_token, '() n e -> b n e', b=b)
            # prepend the cls token to the input
            x = torch.cat([cls_tokens, x], dim=1)
            # add position embedding
            x += self.positions
            return x

## Self Attention
![v3](https://user-images.githubusercontent.com/47969237/141843144-8552c3cb-3fa3-47ca-8081-f5f3db8d257f.jpg)

I study at school 이라는 문장이 있다고 해보자. 각 단어에 대한 임베딩을 구하는 데 있어서 I 라는 단어 하나만 보는 것보다 문장 내에서 다른 단어와 어떤 관계가 있는지 살펴보고 attention score를 구하는 것이 효율적이다. Transformer encoder에 들어가는 query,key,value 가 어떻게 구해지는지를 도식화한 그림이다. 각 q, k , v마다 적용되는 weight matrix 를 곱해주어 단어 별로 q,k v 를 구하게 된다.

I study at school 이라는 문장이 있다고 해보자. 각 단어에 대한 임베딩을 구하는 데 있어서 I 라는 단어 하나만 보는 것보다 문장 내에서 다른 단어와 어떤 관계가 있는지 살펴보고 attention score를 구하는 것이 효율적이다. Transformer encoder에 들어가는 query,key,value 가 어떻게 구해지는지를 도식화한 그림이다. 각 q, k , v마다 적용되는 weight matrix 를 곱해주어 단어 별로 q,k v 를 구하게 된다. -> **Attention layer output**

## Multi-Head Attention
![v4](https://user-images.githubusercontent.com/47969237/141843260-8526c885-e107-4060-b46b-4c427c563388.jpg)

attention head를 여러개 가지는 것을 Multi-head attention이라고 한다.  각각의 attention head는 다른 Q, K, V 가중치 행렬들을 갖고, Q, K, V 행렬들을 결과로 갖는 어텐션을 계산한다. 그리고 그 Z를 모두 모아서(concat) 가중치 행렬과 곱해서 레이어의 최종 출력값을 만든다.

    class MultiHeadAttention(nn.Module):
        def __init__(self, emb_size: int = 512, num_heads: int = 8, dropout: float = 0):
            super().__init__()
            self.emb_size = emb_size
            self.num_heads = num_heads
            self.keys = nn.Linear(emb_size, emb_size)
            self.queries = nn.Linear(emb_size, emb_size)
            self.values = nn.Linear(emb_size, emb_size)
            self.att_drop = nn.Dropout(dropout)
            self.projection = nn.Linear(emb_size, emb_size)
            
        def forward(self, x : Tensor, mask: Tensor = None) -> Tensor:
            # split keys, queries and values in num_heads
            queries = rearrange(self.queries(x), "b n (h d) -> b h n d", h=self.num_heads)
            keys = rearrange(self.keys(x), "b n (h d) -> b h n d", h=self.num_heads)
            values  = rearrange(self.values(x), "b n (h d) -> b h n d", h=self.num_heads)
            # 쿼리,키 연산
            energy = torch.einsum('bhqd, bhkd -> bhqk', queries, keys) # batch, num_heads, query_len, key_len
            if mask is not None:
                fill_value = torch.finfo(torch.float32).min
                energy.mask_fill(~mask, fill_value)
            
            #어텐션 스코어
            scaling = self.emb_size ** (1/2)
            att = F.softmax(energy, dim=-1) / scaling
            att = self.att_drop(att)
            # sum up over the third axis
            out = torch.einsum('bhal, bhlv -> bhav ', att, values)
            out = rearrange(out, "b h n d -> b n (h d)")
            out = self.projection(out)
            return out

## MLP
MHA 이후 진행되는 MLP에 대한 코드이다. Linear, GELU, Dropout, Linear로 구성되어 있으며 임베딩 사이즈의 4배만큼 expansion을 곱해주고 dropout 이후에는 다시 원래 사이즈로 축소한다

    class FeedForwardBlock(nn.Sequential):
        def __init__(self, emb_size: int, expansion: int = 4, drop_p: float = 0.):
            super().__init__(
                nn.Linear(emb_size, expansion * emb_size),
                nn.GELU(),
                nn.Dropout(drop_p),
                nn.Linear(expansion * emb_size, emb_size),
            )

### Residual Connection
트랜스포머 인코더에는 residual connection이 적용된다.
문제 : 레이어를 깊게 쌓았음에도 불구하고 Shallow layer 보다 성능이 안좋다. Optimization 이 어렵다.

Plain layer 에서는 y = f(x) 이므로 NN가 고차원적인 feature vector 로의 Mapping 이라고 봤을 때, 레이어가 쌓일수록 한번에 Mapping 하면서 학습해야 하는 양이 너무 많아져서 학습이 어렵게 된다.
Residual block 에서는 y = f(x) + x , 즉 input x를 더해주게 되며 기존의 x를 보존하면서 추가적인 정보를 학습한다고 이해할 때 학습량이 적어지게 된다.

    class ResidualAdd(nn.Module):
        def __init__(self, fn):
		        super().__init__()
		        self.fn = fn
        
        def forward(self,x,**kwargs):
		        res = x 				
		        x = self.fn(x,kwargs)
		        x += res
		        return x

### Transformer Encoder

    class TransformerEncoderBlock(nn.Sequential):
        def __init__(self,
                     emb_size: int = 768,
                     drop_p: float = 0.,
                     forward_expansion: int = 4,
                     forward_drop_p: float = 0.,
                     ** kwargs):
            super().__init__(
                ResidualAdd(nn.Sequential(
                    nn.LayerNorm(emb_size),
                    MultiHeadAttention(emb_size, **kwargs),
                    nn.Dropout(drop_p)
                )),
                ResidualAdd(nn.Sequential(
                    nn.LayerNorm(emb_size),
                    FeedForwardBlock(
                        emb_size, expansion=forward_expansion, drop_p=forward_drop_p),
                    nn.Dropout(drop_p)
                )
                ))
